In [124]:
import pandas as pd
import numpy as np

veh_df = pd.read_csv('Vehicle_Information.csv', encoding='ISO-8859-1')

## **Vehicle info**

In [125]:
veh_df.head(10)

Accident_Index Age_Band_of_Driver  Age_of_Vehicle  \
0  200401BS00001            26 - 35             3.0   
1  200401BS00002            26 - 35             NaN   
2  200401BS00003            26 - 35             4.0   
3  200401BS00003            66 - 75             NaN   
4  200401BS00004            26 - 35             1.0   
5  200401BS00004            36 - 45            10.0   
6  200401BS00009            26 - 35             NaN   
7  200401BS00010            36 - 45             NaN   
8  200401BS00012            46 - 55             3.0   
9  200401BS00013            26 - 35             4.0   

          Driver_Home_Area_Type  Driver_IMD_Decile  Engine_Capacity_.CC.  \
0                    Urban area                4.0                1588.0   
1                    Urban area                3.0                   NaN   
2  Data missing or out of range                NaN                 998.0   
3  Data missing or out of range                NaN                   NaN   
4                    Urban area                4.0                 124.0   
5  Data missing or out of range                NaN                1781.0   
6                    Urban area                4.0                   NaN   
7                    Urban area                8.0                   NaN   
8  Data missing or out of range                NaN                2685.0   
9                    Urban area                6.0                2300.0   

  Hit_Object_in_Carriageway Hit_Object_off_Carriageway  \
0                      None                       None   
1                      None                       None   
2                      None                       None   
3                      None                       None   
4                      None                       None   
5                      None                       None   
6                      None                       None   
7                      None                       None   
8                      None                       None   
9                      None                       None   

      Journey_Purpose_of_Driver             Junction_Location  ...  \
0  Data missing or out of range  Data missing or out of range  ...   
1  Data missing or out of range  Data missing or out of range  ...   
2  Data missing or out of range  Data missing or out of range  ...   
3  Data missing or out of range  Data missing or out of range  ...   
4  Data missing or out of range  Data missing or out of range  ...   
5  Data missing or out of range  Data missing or out of range  ...   
6  Data missing or out of range  Data missing or out of range  ...   
7  Data missing or out of range  Data missing or out of range  ...   
8  Data missing or out of range  Data missing or out of range  ...   
9  Data missing or out of range  Data missing or out of range  ...   

  Skidding_and_Overturning Towing_and_Articulation  \
0                     None     No tow/articulation   
1                     None     No tow/articulation   
2                     None     No tow/articulation   
3                     None     No tow/articulation   
4                     None     No tow/articulation   
5                     None     No tow/articulation   
6                  Skidded     No tow/articulation   
7                     None     No tow/articulation   
8                     None     No tow/articulation   
9                     None     No tow/articulation   

  Vehicle_Leaving_Carriageway Vehicle_Location.Restricted_Lane  \
0   Did not leave carriageway                              0.0   
1   Did not leave carriageway                              0.0   
2   Did not leave carriageway                              0.0   
3   Did not leave carriageway                              0.0   
4   Did not leave carriageway                              0.0   
5   Did not leave carriageway                              0.0   
6   Did not leave carriageway                              0.0   
7   Did not leave carriage

In [126]:
veh_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2177205 entries, 0 to 2177204
Data columns (total 24 columns):
 #   Column                            Dtype  
---  ------                            -----  
 0   Accident_Index                    object 
 1   Age_Band_of_Driver                object 
 2   Age_of_Vehicle                    float64
 3   Driver_Home_Area_Type             object 
 4   Driver_IMD_Decile                 float64
 5   Engine_Capacity_.CC.              float64
 6   Hit_Object_in_Carriageway         object 
 7   Hit_Object_off_Carriageway        object 
 8   Journey_Purpose_of_Driver         object 
 9   Junction_Location                 object 
 10  make                              object 
 11  model                             object 
 12  Propulsion_Code                   object 
 13  Sex_of_Driver                     object 
 14  Skidding_and_Overturning          object 
 15  Towing_and_Articulation           object 
 16  Vehicle_Leaving_Carriageway       ob

In [127]:
veh_df.describe()

Age_of_Vehicle  Driver_IMD_Decile  Engine_Capacity_.CC.  \
count    1.819056e+06       1.442393e+06          1.911344e+06   
mean     7.108167e+00       5.387559e+00          2.042234e+03   
std      4.725886e+00       2.821651e+00          1.950143e+03   
min      1.000000e+00       1.000000e+00          1.000000e+00   
25%      3.000000e+00       3.000000e+00          1.299000e+03   
50%      7.000000e+00       5.000000e+00          1.598000e+03   
75%      1.000000e+01       8.000000e+00          1.997000e+03   
max      1.110000e+02       1.000000e+01          9.600000e+04   

       Vehicle_Location.Restricted_Lane  Vehicle_Reference          Year  
count                      2.175888e+06       2.177205e+06  2.177205e+06  
mean                       1.073042e-01       1.553405e+00  2.010934e+03  
std                        8.799644e-01       7.752484e-01  3.694375e+00  
min                        0.000000e+00       1.000000e+00  2.004000e+03  
25%                        0.000000e+00       1.000000e+00  2.008000e+03  
50%                        0.000000e+00       1.000000e+00  2.011000e+03  
75%                        0.000000e+00       2.000000e+00  2.014000e+03  
max                        9.000000e+00       9.100000e+01  2.016000e+03

In [128]:
veh_df.shape

(2177205, 24)

In [129]:
veh_df.isnull().sum()

Accident_Index                           0
Age_Band_of_Driver                       0
Age_of_Vehicle                      358149
Driver_Home_Area_Type                    0
Driver_IMD_Decile                   734812
Engine_Capacity_.CC.                265861
Hit_Object_in_Carriageway                0
Hit_Object_off_Carriageway               0
Journey_Purpose_of_Driver                0
Junction_Location                        0
make                                110845
model                               325331
Propulsion_Code                     245843
Sex_of_Driver                            0
Skidding_and_Overturning                 0
Towing_and_Articulation                  0
Vehicle_Leaving_Carriageway              0
Vehicle_Location.Restricted_Lane      1317
Vehicle_Manoeuvre                        0
Vehicle_Reference                        0
Vehicle_Type                             0
Was_Vehicle_Left_Hand_Drive              0
X1st_Point_of_Impact                     0
Year       

## **1) Handling missing values**

##### **Age_of_Vehicle**
Imputing the missing values in the Age_of_Vehicle column by preserving the existing age distribution in the dataset.

We first calculate the probability distribution of non-null Age_of_Vehicle values. This is done by dividing the value counts of each unique age by the total number of non-null ages.

Next, we impute the missing values by randomly sampling from the non-null values using the calculated probability distribution. This ensures that the imputed ages follow the same distribution as the original data.

By using this approach, we maintain the age distribution of the vehicles in the dataset, ensuring that the imputed values do not introduce any bias or imbalance.

In [130]:
# compute the value counts and convert to a DataFrame
value_counts_df = veh_df['Age_of_Vehicle'].value_counts().reset_index().rename(columns={'index': 'Age', 'Age_of_Vehicle': 'Count'})

# save the DataFrame to a CSV file, to check as refernce to confirm imputation has happenned 
value_counts_df.to_csv('age_of_vehicle_counts.csv', index=False)

In [131]:
# Calculate the probability distribution of non-null Age_of_Vehicle values
non_null_age_of_vehicle = veh_df[veh_df['Age_of_Vehicle'].notnull()]['Age_of_Vehicle']
value_counts = non_null_age_of_vehicle.value_counts()
prob_distribution = value_counts / value_counts.sum()

# Impute the missing values using the calculated probability distribution
num_missing_values = veh_df['Age_of_Vehicle'].isnull().sum()
imputed_values = np.random.choice(prob_distribution.index, size=num_missing_values, p=prob_distribution.values)

# Assign the imputed values to the missing values in the Age_of_Vehicle column
veh_df.loc[veh_df['Age_of_Vehicle'].isnull(), 'Age_of_Vehicle'] = imputed_values

checking successful imputation comparing value counts before and after

In [132]:
# compute the value counts and convert to a DataFrame
value_counts_df = veh_df['Age_of_Vehicle'].value_counts().reset_index().rename(columns={'index': 'Age', 'Age_of_Vehicle': 'Count'})

# save the DataFrame to a CSV file
value_counts_df.to_csv('age_of_vehicle_counts.csv2', index=False)

##### **Driver_IMD_Decile**
imputing a large number of missing values (about 34% in this case) using random sampling might not be the best approach, as it could introduce more noise to the dataset. Also, there isn't any strong correlation between the Driver_IMD_Decile and other columns logically, so it would be difficult to accurately estimate the missing values based on other features.

In this case, I will go with replacing the NaN values with -1, which represents "unknown" or "missing data." This will keep the missing values separate from the actual data and allows for further analysis or modeling to take into account the presence of unknown IMD decile values

In [133]:
veh_df['Driver_IMD_Decile'] = veh_df['Driver_IMD_Decile'].astype('str').replace('nan', '-1').astype('category')

##### **Engine_Capacity_.CC.**
fill the missing Engine_Capacity_.CC. values using the average engine capacity for the same make and model in the dataset, given that both make and model values are present for the row with the missing engine capacity.

In [134]:
# Step 1: Calculate the mean engine capacity for each make and model
mean_engine_capacity = veh_df.groupby(['make', 'model'])['Engine_Capacity_.CC.'].mean().reset_index()

# Step 2: Define a function to replace NaN engine capacity with the mean engine capacity for the same make and model
def fill_engine_capacity(row, mean_engine_capacity):
    if pd.isnull(row['Engine_Capacity_.CC.']):
        make = row['make']
        model = row['model']
        if pd.notnull(make) and pd.notnull(model):
            mean_capacity = mean_engine_capacity.loc[(mean_engine_capacity['make'] == make) & (mean_engine_capacity['model'] == model), 'Engine_Capacity_.CC.']
            if not mean_capacity.empty:
                return mean_capacity.values[0]
    return row['Engine_Capacity_.CC.']

# Step 3: Apply the function to the entire dataset
veh_df['Engine_Capacity_.CC.'] = veh_df.apply(lambda row: fill_engine_capacity(row, mean_engine_capacity), axis=1)


In [135]:
# filling in any values that couldnt be changed due to other column values not being present to -1 as to be unknown
veh_df['Engine_Capacity_.CC.'].fillna(-1, inplace=True)

##### **Make and Model**
Since each car model is specific to one make, we can create a dictionary mapping each model to its corresponding make. Then, we can use this dictionary to fill in the missing make values when the model is available.

In [136]:
# Step 1: Create a dictionary mapping each model to its corresponding make
make_model_dict = veh_df.loc[veh_df['make'].notnull() & veh_df['model'].notnull()].set_index('model')['make'].to_dict()

# Step 2: Create a function to fill in the missing make values
def fill_missing_make(row):
    make, model = row['make'], row['model']
    if pd.isna(make) and model in make_model_dict:
        return make_model_dict[model]
    return make

# Step 3: Apply the function to the entire dataset
veh_df['make'] = veh_df.apply(fill_missing_make, axis=1)

In [137]:
veh_df['make'].isnull().sum()

110845

In [138]:
missing_both = veh_df[(veh_df['make'].isnull()) & (veh_df['model'].isnull())].shape[0]
print(f"Number of rows missing both make and model: {missing_both}")


Number of rows missing both make and model: 110845


After running that code and seeing no change in missing make values, it is clear that in all rows where make is missing, model is also missing. So another method will have to be used. What this does mean though is that since there are 110845 missing make values and 325331 missing model values, is that there are 214,486 rows where make is present and model is missing, therefore i would be happy to substitute the model in these rows for the mode of each make

In [139]:
# Step 1: Create a dictionary mapping each 'make' to its mode 'model'
make_model_mode = veh_df[veh_df['model'].notnull()].groupby('make')['model'].agg(pd.Series.mode).to_dict()

# Step 2: Create a function that takes a row as input and returns the mode 'model' for the given 'make' if the 'model' value is missing
def fill_model(row):
    if pd.isnull(row['model']) and not pd.isnull(row['make']):
        return make_model_mode.get(row['make'], np.nan)
    else:
        return row['model']

# Step 3: Apply the function to the DataFrame to update the 'model' column
veh_df['model'] = veh_df.apply(fill_model, axis=1)


In [140]:
veh_df['model'].isnull().sum()

112449

In [141]:
rows_with_make_enginecap_no_model = veh_df[(veh_df['make'].notnull()) & (veh_df['model'].isnull())].shape[0]
print(rows_with_make_enginecap_no_model)

1604


In [142]:
# Select the 1604 rows where 'make' is missing but 'model' is present
missing_make_df = veh_df[(veh_df['make'].isnull()) & (veh_df['model'].notnull())]

# Fill in the missing 'make' values using the make_model_dict
missing_make_df['make'] = missing_make_df['model'].map(lambda x: make_model_dict.get(x))

# Update the original dataset with the filled-in 'make' values
veh_df.update(missing_make_df)

changing remaining unknown make and model values to -1, as they occur in the same rows as each other and we cant accurately predict what either one of them will be without the other. 

In [143]:
veh_df['make'] = veh_df['make'].fillna('-1').astype(str)
veh_df['model'] = veh_df['model'].fillna('-1').astype(str)
veh_df['make'] = veh_df['make'].astype('category')
veh_df['model'] = veh_df['model'].astype('category')

##### **Propulsion_Code**
Converting unknowns to -1 and changinb to category type

In [144]:
veh_df['Propulsion_Code'] = veh_df['Propulsion_Code'].fillna('Undefined').astype('category')

##### **Vehicle_Location.Restricted_Lane**
same method as propulsion code

In [145]:
veh_df['Vehicle_Location.Restricted_Lane'] = veh_df['Vehicle_Location.Restricted_Lane'].astype('str').replace('nan', '-1').astype('category')


**Checking imputation has worked for all**

In [146]:
veh_df.isnull().sum()

Accident_Index                      0
Age_Band_of_Driver                  0
Age_of_Vehicle                      0
Driver_Home_Area_Type               0
Driver_IMD_Decile                   0
Engine_Capacity_.CC.                0
Hit_Object_in_Carriageway           0
Hit_Object_off_Carriageway          0
Journey_Purpose_of_Driver           0
Junction_Location                   0
make                                0
model                               0
Propulsion_Code                     0
Sex_of_Driver                       0
Skidding_and_Overturning            0
Towing_and_Articulation             0
Vehicle_Leaving_Carriageway         0
Vehicle_Location.Restricted_Lane    0
Vehicle_Manoeuvre                   0
Vehicle_Reference                   0
Vehicle_Type                        0
Was_Vehicle_Left_Hand_Drive         0
X1st_Point_of_Impact                0
Year                                0
dtype: int64

#### **Inspecting other columns where missing values may not be 'nan'**
##### **Age_Band_of_Driver**

After closer inspection of Age band of drivers column it we can see there is a large amount of data as 'Data missing or out of range', so ti handle this, in order to keep these rows of data usaeable we can change the missing values by evenly distrubutinh the other age bands in teh ratios they appear in the data set so it doesnt skew this data any direction. 

In [147]:
veh_df['Age_Band_of_Driver'].value_counts()

26 - 35                         450531
36 - 45                         435686
46 - 55                         348762
21 - 25                         238765
56 - 65                         206181
16 - 20                         175874
Data missing or out of range    171052
66 - 75                          91454
Over 75                          54236
11 - 15                           3655
6 - 10                             884
0 - 5                              125
Name: Age_Band_of_Driver, dtype: int64

In [148]:
# Step 1: Calculate the probability distribution of the existing age bands
age_band_counts = veh_df[veh_df['Age_Band_of_Driver'] != 'Data missing or out of range']['Age_Band_of_Driver'].value_counts()
age_band_probs = age_band_counts / age_band_counts.sum()

# Step 2: Replace 'Data missing or out of range' with a randomly chosen age band
mask = veh_df['Age_Band_of_Driver'] == 'Data missing or out of range'
missing_age_band_count = mask.sum()

random_age_bands = np.random.choice(age_band_probs.index, size=missing_age_band_count, p=age_band_probs.values)
veh_df.loc[mask, 'Age_Band_of_Driver'] = random_age_bands

# Check the updated value counts
print(veh_df['Age_Band_of_Driver'].value_counts())

26 - 35    488989
36 - 45    472637
46 - 55    378489
21 - 25    259368
56 - 65    223549
16 - 20    190868
66 - 75     99339
Over 75     58913
11 - 15      3967
6 - 10        951
0 - 5         135
Name: Age_Band_of_Driver, dtype: int64


##### **Driver_Hone_Area_Type**

In [149]:
veh_df['Driver_Home_Area_Type'].value_counts()

Urban area                      1436598
Data missing or out of range     334344
Rural                            232360
Small town                       173903
Name: Driver_Home_Area_Type, dtype: int64

Just keeping as is, as replacing Data missing here would skew the data too much

##### **Hit_Object_in_Carriageway**

In [150]:
veh_df['Hit_Object_in_Carriageway'].value_counts()

None                                2087824
Kerb                                  34354
Parked vehicle                        27401
Bollard or refuge                      9981
Other object                           5715
Any animal (except ridden horse)       2683
Central island of roundabout           2190
Open door of vehicle                   1919
Bridge (side)                          1441
Data missing or out of range           1302
Previous accident                      1023
Road works                              957
Bridge (roof)                           415
Name: Hit_Object_in_Carriageway, dtype: int64

Keeping data as it is

In [151]:
veh_df['Hit_Object_off_Carriageway'].value_counts()

None                             1989741
Other permanent object             56640
Tree                               29161
Entered ditch                      18301
Road sign or traffic signal        17394
Central crash barrier              16538
Near/Offside crash barrier         15835
Lamp post                          15293
Wall or fence                      10423
Telegraph or electricity pole       5828
Bus stop or bus shelter             1300
Data missing or out of range         533
Submerged in water                   218
Name: Hit_Object_off_Carriageway, dtype: int64

In [152]:
veh_df['Journey_Purpose_of_Driver'].value_counts()

Not known                       834626
Other/Not known (2005-10)       555726
Journey as part of work         391713
Commuting to/from work          202525
Data missing or out of range    133560
Other                            32965
Taking pupil to/from school      21792
Pupil riding to/from school       4298
Name: Journey_Purpose_of_Driver, dtype: int64

In [153]:
veh_df['Junction_Location'].value_counts()

Not at or within 20 metres of junction                         808108
Approaching junction or waiting/parked at junction approach    476804
Mid Junction - on roundabout or on main road                   423304
Data missing or out of range                                   122612
Cleared junction or waiting/parked at junction exit            114061
Entering main road                                              89797
Entering roundabout                                             54546
Leaving main road                                               51094
Leaving roundabout                                              28965
Entering from slip road                                          7914
Name: Junction_Location, dtype: int64

In [154]:
veh_df['Propulsion_Code'].value_counts()

Petrol                 1143097
Heavy oil               775829
Undefined               245843
Hybrid electric           9172
Gas/Bi-fuel               1642
Petrol/Gas (LPG)           664
Electric                   618
Gas                        175
Electric diesel            143
New fuel technology         11
Gas Diesel                   7
Steam                        3
Fuel cells                   1
Name: Propulsion_Code, dtype: int64

In [155]:
veh_df['Sex_of_Driver'].value_counts()

Male                            1468081
Female                           633005
Not known                         76051
Data missing or out of range         68
Name: Sex_of_Driver, dtype: int64

In [156]:
veh_df['Skidding_and_Overturning'].value_counts()

None                            1898208
Skidded                          199825
Skidded and overturned            42339
Overturned                        34100
Jackknifed                         1127
Data missing or out of range       1015
Jackknifed and overturned           591
Name: Skidding_and_Overturning, dtype: int64

In [157]:
veh_df['Vehicle_Type'].value_counts()

Car                                      1528628
Van / Goods 3.5 tonnes mgw or under       117427
109                                        82920
Bus or coach (17 or more pass seats)       76757
Motorcycle over 500cc                      71472
Motorcycle 125cc and under                 61600
Goods 7.5 tonnes mgw and over              55426
Taxi/Private hire car                      43781
Pedal cycle                                38904
Motorcycle 50cc and under                  22415
Motorcycle over 125cc and up to 500cc      20960
Goods over 3.5t. and under 7.5t            18236
Other vehicle                              13994
106                                         7568
Agricultural vehicle                        6018
Minibus (8 - 16 passenger seats)            5900
Goods vehicle - unknown weight              1876
108                                         1334
Motorcycle - unknown cc                      741
Mobility scooter                             502
Data missing or out 

In [158]:
veh_df['Vehicle_Type'].value_counts()

Car                                      1528628
Van / Goods 3.5 tonnes mgw or under       117427
109                                        82920
Bus or coach (17 or more pass seats)       76757
Motorcycle over 500cc                      71472
Motorcycle 125cc and under                 61600
Goods 7.5 tonnes mgw and over              55426
Taxi/Private hire car                      43781
Pedal cycle                                38904
Motorcycle 50cc and under                  22415
Motorcycle over 125cc and up to 500cc      20960
Goods over 3.5t. and under 7.5t            18236
Other vehicle                              13994
106                                         7568
Agricultural vehicle                        6018
Minibus (8 - 16 passenger seats)            5900
Goods vehicle - unknown weight              1876
108                                         1334
Motorcycle - unknown cc                      741
Mobility scooter                             502
Data missing or out 

As we can see here, some values are a labeled as their code value instead of the name value, so we will change them in accordance with the data info we have by creating a dictionary of codes and names to map.

In [159]:
# Step 1: Create a dictionary to map vehicle type codes to names
vehicle_type_mapping = {
    1: 'Pedal cycle',
    2: 'Motorcycle 50cc and under',
    3: 'Motorcycle 125cc and under',
    4: 'Motorcycle over 125cc and up to 500cc',
    5: 'Motorcycle over 500cc',
    8: 'Taxi/Private hire car',
    9: 'Car',
    10: 'Minibus (8 - 16 passenger seats)',
    11: 'Bus or coach (17 or more pass seats)',
    16: 'Ridden horse',
    17: 'Agricultural vehicle',
    18: 'Tram',
    19: 'Van / Goods 3.5 tonnes mgw or under',
    20: 'Goods over 3.5t. and under 7.5t',
    21: 'Goods 7.5 tonnes mgw and over',
    22: 'Mobility scooter',
    23: 'Electric motorcycle',
    90: 'Other vehicle',
    97: 'Motorcycle - unknown cc',
    98: 'Goods vehicle - unknown weight',
    99: 'Unknown vehicle type (self rep only)',
    103: 'Motorcycle - Scooter (1979-1998)',
    104: 'Motorcycle (1979-1998)',
    105: 'Motorcycle - Combination (1979-1998)',
    106: 'Motorcycle over 125cc (1999-2004)',
    108: 'Taxi (excluding private hire cars) (1979-2004)',
    109: 'Car (including private hire cars) (1979-2004)',
    110: 'Minibus/Motor caravan (1979-1998)',
    113: 'Goods over 3.5 tonnes (1979-1998)',
    -1: 'Data missing or out of range'
}

# Step 2: Replace numeric values in the 'Vehicle_Type' column with their corresponding names
veh_df['Vehicle_Type'] = veh_df['Vehicle_Type'].apply(lambda x: vehicle_type_mapping[int(x)] if str(x).isnumeric() else x)

In [160]:
veh_df['Was_Vehicle_Left_Hand_Drive'].value_counts()

No                              2045375
Data missing or out of range     127943
Yes                                3887
Name: Was_Vehicle_Left_Hand_Drive, dtype: int64

## **2) Handling data types**

In [161]:
veh_df.dtypes

Accident_Index                        object
Age_Band_of_Driver                    object
Age_of_Vehicle                       float64
Driver_Home_Area_Type                 object
Driver_IMD_Decile                   category
Engine_Capacity_.CC.                 float64
Hit_Object_in_Carriageway             object
Hit_Object_off_Carriageway            object
Journey_Purpose_of_Driver             object
Junction_Location                     object
make                                category
model                               category
Propulsion_Code                     category
Sex_of_Driver                         object
Skidding_and_Overturning              object
Towing_and_Articulation               object
Vehicle_Leaving_Carriageway           object
Vehicle_Location.Restricted_Lane    category
Vehicle_Manoeuvre                     object
Vehicle_Reference                      int64
Vehicle_Type                          object
Was_Vehicle_Left_Hand_Drive           object
X1st_Point

In [162]:
# Convert object columns to category
category_columns = [
    'Age_Band_of_Driver',
    'Driver_Home_Area_Type',
    'Hit_Object_in_Carriageway',
    'Hit_Object_off_Carriageway',
    'Journey_Purpose_of_Driver',
    'Junction_Location',
    'Sex_of_Driver',
    'Skidding_and_Overturning',
    'Towing_and_Articulation',
    'Vehicle_Leaving_Carriageway',
    'Vehicle_Manoeuvre',
    'Vehicle_Type',
    'Was_Vehicle_Left_Hand_Drive',
    'X1st_Point_of_Impact'
]

for col in category_columns:
    veh_df[col] = veh_df[col].astype('category')

# Convert float columns to the appropriate data type
veh_df['Age_of_Vehicle'] = veh_df['Age_of_Vehicle'].fillna(-1).astype('int32')
veh_df['Engine_Capacity_.CC.'] = veh_df['Engine_Capacity_.CC.'].fillna(-1).astype('int32')

# Convert integer columns to the appropriate data type
veh_df['Vehicle_Reference'] = veh_df['Vehicle_Reference'].astype('int32')
veh_df['Year'] = veh_df['Year'].astype('int32')

# Check the updated data types
print(veh_df.dtypes)


Accident_Index                        object
Age_Band_of_Driver                  category
Age_of_Vehicle                         int32
Driver_Home_Area_Type               category
Driver_IMD_Decile                   category
Engine_Capacity_.CC.                   int32
Hit_Object_in_Carriageway           category
Hit_Object_off_Carriageway          category
Journey_Purpose_of_Driver           category
Junction_Location                   category
make                                category
model                               category
Propulsion_Code                     category
Sex_of_Driver                       category
Skidding_and_Overturning            category
Towing_and_Articulation             category
Vehicle_Leaving_Carriageway         category
Vehicle_Location.Restricted_Lane    category
Vehicle_Manoeuvre                   category
Vehicle_Reference                      int32
Vehicle_Type                        category
Was_Vehicle_Left_Hand_Drive         category
X1st_Point

## **Feature Engineering**

creating new column for more general vehicle type

In [163]:
def map_general_vehicle_type(specific_type):
    if specific_type in ["Taxi/Private hire car", "Car", "Taxi (excluding private hire cars) (1979-2004)", "Car (including private hire cars) (1979-2004)"]:
        return "Car"
    elif specific_type in ["Motorcycle 50cc and under", "Motorcycle 125cc and under", "Motorcycle over 125cc and up to 500cc", "Motorcycle over 500cc", "Motorcycle - unknown cc", "Motorcycle - Scooter (1979-1998)", "Motorcycle (1979-1998)", "Motorcycle - Combination (1979-1998)", "Motorcycle over 125cc (1999-2004)", "Electric motorcycle"]:
        return "Motorcycle"
    elif specific_type in ["Minibus (8 - 16 passenger seats)", "Bus or coach (17 or more pass seats)", "Minibus/Motor caravan (1979-1998)"]:
        return "Bus"
    elif specific_type in ["Van / Goods 3.5 tonnes mgw or under", "Goods vehicle - unknown weight"]:
        return "Van"
    elif specific_type in ["Goods over 3.5t. and under 7.5t", "Goods 7.5 tonnes mgw and over", "Goods over 3.5 tonnes (1979-1998)"]:
        return "Lorry"
    elif specific_type == "Pedal cycle":
        return "Bicycle"
    else:
        return "Other"

# Apply the mapping function to create the new 'General_Vehicle_Type' column
veh_df['General_Vehicle_Type'] = veh_df['Vehicle_Type'].apply(map_general_vehicle_type)

veh_df['General_Vehicle_Type'].value_counts()

Car           1656663
Motorcycle     184834
Van            119303
Bus             82657
Lorry           73662
Bicycle         38904
Other           21182
Name: General_Vehicle_Type, dtype: int64

In [164]:
veh_df.head()

Accident_Index Age_Band_of_Driver  Age_of_Vehicle  \
0  200401BS00001            26 - 35               3   
1  200401BS00002            26 - 35               4   
2  200401BS00003            26 - 35               4   
3  200401BS00003            66 - 75              11   
4  200401BS00004            26 - 35               1   

          Driver_Home_Area_Type Driver_IMD_Decile  Engine_Capacity_.CC.  \
0                    Urban area               4.0                  1588   
1                    Urban area               3.0                   122   
2  Data missing or out of range                -1                   998   
3  Data missing or out of range                -1                  2402   
4                    Urban area               4.0                   124   

  Hit_Object_in_Carriageway Hit_Object_off_Carriageway  \
0                      None                       None   
1                      None                       None   
2                      None                       None   
3                      None                       None   
4                      None                       None   

      Journey_Purpose_of_Driver             Junction_Location  ...  \
0  Data missing or out of range  Data missing or out of range  ...   
1  Data missing or out of range  Data missing or out of range  ...   
2  Data missing or out of range  Data missing or out of range  ...   
3  Data missing or out of range  Data missing or out of range  ...   
4  Data missing or out of range  Data missing or out of range  ...   

  Towing_and_Articulation Vehicle_Leaving_Carriageway  \
0     No tow/articulation   Did not leave carriageway   
1     No tow/articulation   Did not leave carriageway   
2     No tow/articulation   Did not leave carriageway   
3     No tow/articulation   Did not leave carriageway   
4     No tow/articulation   Did not leave carriageway   

  Vehicle_Location.Restricted_Lane  Vehicle_Manoeuvre Vehicle_Reference  \
0                              0.0  Going ahead other                 2   
1                              0.0  Going ahead other                 1   
2                              0.0      Turning right                 1   
3                              0.0  Going ahead other                 2   
4                              0.0  Going ahead other                 1   

                                    Vehicle_Type  \
0  Car (including private hire cars) (1979-2004)   
1  Car (including private hire cars) (1979-2004)   
2  Car (including private hire cars) (1979-2004)   
3  Car (including private hire cars) (1979-2004)   
4                     Motorcycle 125cc and under   

    Was_Vehicle_Left_Hand_Drive X1st_Point_of_Impact  Year  \
0  Data missing or out of range                Front  2004   
1  Data missing or out of range                Front  2004   
2  Data missing or out of range                Front  2004   
3  Data missing or out of range                Front  2004   
4  Data missing or out of range                Front  2004   

   General_Vehicle_Type  
0                   Car  
1                   Car  
2                   Car  
3                   Car  
4            Motorcycle  

[5 rows x 25 columns]

In [166]:
veh_df.to_csv('clean_veh_info.csv')